### PR TITLE
Added support for raspberrypi pico-w

### PR DESCRIPTION
**Pull request by curuvar**
## Summary

Added support for RaspberryPi Pico W

## Impact

Adding support for the Pico W required changes to the networking software.  The existing BCMF chips use SDIO to communicate with the processor.  The 43439 in the Pi W uses gSPI.  In order to use as much common code as possible, I implemented bcmf_gspi.c to parallel bcmf_spio.c. I also added a file bcmf_interface.c that server as an abstraction layer between the two.

Most files that used "struct bcmf_sdio_dev_s sbus" were changed to use "bcmf_interface_def_t ibus".   The "sbus" name was kept in places specific to SDIO, where "gbus" is used in places specific to gSPI.

## Testing

Have tested the Pico-W in general and the Pico-w WiFi access using telnet, telnetd, ping and renew.

I do not have the hardware to test the existing SPIO code.

## Known Bugs

DNS queries to don't currently work.

### CODE REVIEW

File: arch/arm/src/Makefile [hunk 1]
```diff
@@ -157,7 +157,7 @@ endef
 nuttx$(EXEEXT): $(HEAD_OBJ) board$(DELIM)libboard$(LIBEXT) $(ARCHSCRIPT)
 	$(Q) echo "LD: nuttx"
 ifneq ($(CONFIG_ALLSYMS),y)
-	$(Q) $(LD) --entry=__start $(LDFLAGS) $(LIBPATHS) $(EXTRA_LIBPATHS) \
+	$(LD) --entry=__start $(LDFLAGS) $(LIBPATHS) $(EXTRA_LIBPATHS) \
```
Comment (xiaoxiang781216) [inline arch/arm/src/Makefile:160]: why change this?

File: arch/arm/src/rp2040/Kconfig [hunk 2]
```diff
@@ -685,6 +685,8 @@ config RP2040_BOARD_HAS_WS2812
 	bool "Has ws2812 pixels"
 	default n
 	depends on WS2812
+  select WS2812_NON_SPI_DRIVER
```
Comment (xiaoxiang781216) [inline arch/arm/src/rp2040/Kconfig:688]: tab

File: arch/arm/src/rp2040/Kconfig [hunk 3]
```diff
@@ -685,6 +685,8 @@ config RP2040_BOARD_HAS_WS2812
...
 	---help---
 		See the Board Selection menu to configure the pins used
 		by ws2812.
+
```
Comment (xiaoxiang781216) [inline arch/arm/src/rp2040/Kconfig:692]: revert the change

File: libs/libc/netdb/lib_dnsquery.c [hunk 4]
```diff
@@ -709,6 +709,14 @@ int dns_query(FAR const char *hostname, FAR union dns_addr_u *addr,
   FAR struct dns_query_s query;
   int ret;
 
+  /* Has the DNS client been properly initialized? */
+
+  if (!dns_initialize())
```
Comment (xiaoxiang781216) [inline libs/libc/netdb/lib_dnsquery.c:714]: we have a better patch to fix this problem. @zhhyu7 please provide the patch.

File: net/arp/arp_out.c [hunk 5]
```diff
@@ -260,7 +260,10 @@ void arp_out(FAR struct net_driver_s *dev)
   ret = arp_find(ipaddr, &ethaddr);
   if (ret < 0)
     {
-      ninfo("ARP request for IP %08lx\n", (unsigned long)ipaddr);
+      ninfo("ARP request for IP %ld.%ld.%ld.%ld\n", ip4_addr1(ipaddr),
```
Comment (xiaoxiang781216) [inline net/arp/arp_out.c:263]: let's move to other patch

File: net/arp/arp_arpin.c [hunk 6]
```diff
@@ -108,7 +108,10 @@ void arp_arpin(FAR struct net_driver_s *dev)
   switch (arp->ah_opcode)
     {
       case HTONS(ARP_REQUEST):
-        ninfo("ARP request for IP %04lx\n", (unsigned long)ipaddr);
+        ninfo("ARP request for IP %ld.%ld.%ld.%ld\n", ip4_addr1(ipaddr),
```
Comment (xiaoxiang781216) [inline net/arp/arp_arpin.c:111]: move to other patch

File: include/nuttx/wireless/ieee80211/bcmf_gspi.h [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __INCLUDE_NUTTX_WIRELESS_IEEE80211_BCMF_GSPI_H
+#define __INCLUDE_NUTTX_WIRELESS_IEEE80211_BCMF_GSPI_H
```
Comment (xiaoxiang781216) [inline include/nuttx/wireless/ieee80211/bcmf_gspi.h:22]: could we split the bcmf related change to new PR?

File: libs/libc/netdb/lib_dnsquery.c (hunk 4 above)
Comment (xiaoxiang781216) [reply libs/libc/netdb/lib_dnsquery.c:714]: @curuvar please review the patch here: https://github.com/apache/incubator-nuttx/pull/6841

File: drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c [hunk 8]
```diff
@@ -252,7 +252,7 @@ static void bcmf_receive(FAR struct bcmf_dev_s *priv)
       priv->bc_dev.d_buf = frame->data;
       priv->bc_dev.d_len = frame->len - (frame->data - frame->base);
 
-      wlinfo("Got frame %p %d\n", frame, priv->bc_dev.d_len);
+      /* wlinfo("Got frame %p %d\n", frame, priv->bc_dev.d_len); */
```
Comment (davids5) [inline drivers/wireless/ieee80211/bcm43xxx/bcmf_netdev.c:255]: #if 0 would be preferred of commenting out

**Review by davids5 — CHANGES_REQUESTED** [review]
@curuvar - Very Nice!
See inline. I noticed wrapped `#define OK` added. Would it be better to not redefine but include the header?

File: arch/arm/src/Makefile (hunk 1 above)
Comment (curuvar) [reply arch/arm/src/Makefile:160]: It was done for testing as I had some weird linking issue.  I'll revert it.

File: include/nuttx/wireless/ieee80211/bcmf_gspi.h (hunk 7 above)
Comment (curuvar) [reply include/nuttx/wireless/ieee80211/bcmf_gspi.h:22]: Sure.  

To be clear I will pull all the actual Pico-W related changes in the "boards" and "arch/arm" directories and resubmit the PR as "adding gSPI support to bcmf".  I will submit a separate PR for "adding pico-w".

Does this match your intent?

Comment (xiaoxiang781216) [reply include/nuttx/wireless/ieee80211/bcmf_gspi.h:22]: Yes, the split looks reasonable.

**Review by btashton — CHANGES_REQUESTED** [review]
There are two binary firmware blobs that cannot be checked into this repo.

File: arch/arm/src/Makefile (hunk 1 above)
Comment (spiriou) [inline arch/arm/src/Makefile:160]: Revert this change?
